File: hello/scala/src/main/scala/polyglot/hello/Greeter.scala
package polyglot.hello

object Greeter {

    def hello( name: String ): String = hello( Option( name ) )

    def hello( name: Option[ String ] = None ) = name match {
        case Some( str ) => "Hello, " + str + "!"
        case None => "Hello!"
    }

    def main( args: Array[ String ] ) = {
        println( hello( args.headOption ) )
    }
    
}

File: hello/java/src/main/java/polyglot/hello/Greeter.java
package polyglot.hello;

import java.util.Objects;

public final class Greeter {

    public static String hello() {
        return hello( null );
    }
    
    public static String hello( final String name ) {
        return name == null ? "Hello!" : "Hello, " + name + "!";
    }
    
    public static void main( final String[] args ) {
        System.out.println( hello( args.length == 1 ? args[ 0 ] : null ) );
    }

}
